File: lib/Mach1DecodeEmscripten.cpp
#include <emscripten/bind.h>

#include "Mach1DecodeCAPI.h"
#include "external/cpp/Mach1Decode.cpp"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(Mach1Decode) {
    enum_<Mach1PlatformType>("Mach1PlatformType")
        .value("Mach1PlatformDefault", Mach1PlatformDefault)
        .value("Mach1PlatformUnity", Mach1PlatformUnity)
        .value("Mach1PlatformUE", Mach1PlatformUE)
        .value("Mach1PlatformOfEasyCam", Mach1PlatformOfEasyCam)
        .value("Mach1PlatformAndroid", Mach1PlatformAndroid)
        .value("Mach1PlatformiOSPortrait", Mach1PlatformiOSPortrait)
        .value("Mach1PlatformiOSLandscape", Mach1PlatformiOSLandscape)
        ;
    enum_<Mach1DecodeAlgoType>("Mach1DecodeAlgoType")
        .value("Mach1DecodeAlgoSpatial", Mach1DecodeAlgoSpatial)
        .value("Mach1DecodeAlgoAltSpatial", Mach1DecodeAlgoAltSpatial)
        .value("Mach1DecodeAlgoHorizon", Mach1DecodeAlgoHorizon)
        .value("Mach1DecodeAlgoHorizonPairs", Mach1DecodeAlgoHorizonPairs)
        .value("Mach1DecodeAlgoSpatialPairs", Mach1DecodeAlgoSpatialPairs)
        ;

	register_vector<float>("VectorFloat");

    value_object<Mach1Point3D>("Mach1Point3D")
		.field("x", &Mach1Point3D::x)
		.field("y", &Mach1Point3D::y)
		.field("z", &Mach1Point3D::z)
        ;

    class_<Mach1Decode>("Mach1Decode")
        .constructor<>()
        .function("setPlatformType", &Mach1Decode::setPlatformType)
        .function("setDecodeAlgoType", &Mach1Decode::setDecodeAlgoType)
        .function("decode", &Mach1Decode::decode)
        .function("setFilterSpeed", &Mach1Decode::setFilterSpeed)
        .function("beginBuffer", &Mach1Decode::beginBuffer)
        .function("endBuffer", &Mach1Decode::endBuffer)
        .function("getCurrentTime", &Mach1Decode::getCurrentTime)
        .function("getCurrentAngle", &Mach1Decode::getCurrentAngle)
        .function("getLog", &Mach1Decode::getLog)
        ;
}

File: lib/Mach1DecodeEmscripten.js
Module['onInited'] = function() {
};

Module['onRuntimeInitialized'] = function() {
	console.log('loaded Mach1Decode');

	Module['Mach1Decode'] = function() {
		this.Mach1PlatformType = {
			"Mach1PlatformDefault": (Module).Mach1PlatformType.Mach1PlatformDefault,
			"Mach1PlatformUnity": (Module).Mach1PlatformType.Mach1PlatformUnity,
			"Mach1PlatformUE": (Module).Mach1PlatformType.Mach1PlatformUE,
			"Mach1PlatformOfEasyCam": (Module).Mach1PlatformType.Mach1PlatformOfEasyCam,
			"Mach1PlatformAndroid": (Module).Mach1PlatformType.Mach1PlatformAndroid,
			"Mach1PlatformiOS": (Module).Mach1PlatformType.Mach1PlatformiOS,
		};

		this.Mach1DecodeAlgoType = {
			"Mach1DecodeAlgoSpatial": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial,
			"Mach1DecodeAlgoAltSpatial": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoAltSpatial,
			"Mach1DecodeAlgoHorizon": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizon,
			"Mach1DecodeAlgoHorizonPairs": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizonPairs,
			"Mach1DecodeAlgoSpatialPairs": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatialPairs,
		};

		this.delete = function() {
			if (this._m1obj) {
				this._m1obj.delete();
				this._m1obj = null;
			}
		};

		this.setPlatformType = function(platform) {
			if (this._m1obj) {
				this._m1obj.setPlatformType(platform);
			}
		};

		this.setDecodeAlgoType = function(decodeAlgo) {
			if (this._m1obj) {
				this._m1obj.setPlatformType(decodeAlgo);
			}
		};

		this.setFilterSpeed = function(filterSpeed) {
			if (this._m1obj) {
				this._m1obj.setFilterSpeed(filterSpeed);
			}
		};

		this.getCurrentAngle = function() {
			if (this._m1obj) {
				return this._m1obj.getCurrentAngle();
			}
		};

		this.decode = function(yaw, pitch, roll, bufferSize, sampleIndex) {
			if (this._m1obj) {
				var v = this._m1obj.decode(yaw, pitch, roll, bufferSize ? bufferSize : 0, sampleIndex ? sampleIndex : 0);
				var a = [];
				for (var i = 0, c = v.size(); i < c; i++) {
					a[i] = v.get(i);
				}
				v.delete();
				return a;
			}
		};
		
		this.beginBuffer = function() {
			if (this._m1obj) {
				this._m1obj.beginBuffer();
			}
		};

		this.endBuffer = function() {
			if (this._m1obj) {
				this._m1obj.endBuffer();
			}
		};    
		
		this._m1obj = new (Module).Mach1DecodeInternal();
	};
	
	Module['onInited']();
};
